### PR TITLE
feat(web): touchpath segment objects 🐵

### PR DESCRIPTION
**Pull request by jahorton**
Continuing from #7058, this PR seeks to construct & publish "proper" Segment objects to the `TrackedPath` object that may be used for later stages of gesture recognition.  These are synthesized by examining the sequence of "subsegments" produced by the segmentation algorithm and recombining them as appropriate based on a few common patterns and rules.

This implements "stage one" from [our "Gesture Recognition" design doc](https://docs.google.com/document/d/1i7WI7iGD0FI2nGx7T65Yu32LgLCaGfIaZN3EPMhlTog/edit?usp=sharing).

Additionally, this PR's changes seek to provide a better, more long-term-maintenance-friendly set of debugging functions, logging methods, etc - rather than maintaining the slew of raw `console.log()` statements its predecessor had.  Accordingly, any debug-logging statements are no longer accessible except through user interaction via the developer console.

------

While I could see there being a desire to split this one into pieces... I'm a little stumped at the moment on how to approach that well - the most-easily separable changes are required by those most directly connected to already-completed work.  Apologies that it grew to this point before being submitted for review, though I'm still glad it wasn't as large as I feared.

------

While user tests are _technically_ possible to write, they'd require parsing the output logs to verify success - and that's not the simplest thing to ask of our user testers.  (I actually started to write a user test or two and then realized just how complex the "verify" instructions were going to be.)

As I plan to follow this up a PR or two for unit-testing, and as this part of the module is meant to be quite "internal" in the grand scheme of things...

@keymanapp-test-bot skip

See #7157 for tests.

### CODE REVIEW

File: common/web/gesture-recognizer/src/trackedPath.ts [hunk 1]
```diff
@@ -29,10 +29,22 @@ namespace com.keyman.osk {
...
    * `'complete'`: the touchpoint is no longer active; a touch-end has been observed.
    *   - Provides no parameters.
+   *   - Will be the last event raised by its instance, after any final 'segmentation'
+   *     events.
    *
    * `'invalidated'`: the touchpoint is no longer active; the path has crossed
    * gesture-recognition boundaries and is no longer considered valid.
    *   - Provides no parameters.
+   *   - Will precede recognition Promise fulfillment on the `Segment` provided by
+   *     the most recently-preceding 'segmentation' event.
```
Comment (jahorton) [inline common/web/gesture-recognizer/src/trackedPath.ts:39]: And shortly after this change, I realized that order doesn't matter, as `Promise` fulfillment and rejection are always async.  No need to make a distinction.

File: common/web/gesture-recognizer/src/trackedPath.ts [hunk 2]
```diff
@@ -100,17 +122,23 @@ namespace com.keyman.osk {
...
-      } else {
+      }
+
+      this.segmenter.close();
+
+      // If not cancelling, signal completion after finishing segments.
+      if(!cancel) {
         this.emit('complete');
       }
+
```
Comment (jahorton) [inline common/web/gesture-recognizer/src/trackedPath.ts:141]: So, this can be simplified.

**Review by mcdurdin — APPROVED** [review]
The code LGTM. Not much I can say on the algorithm without spending a very long time on it -- it seems like there is a lot of complexity and that concerns me a little because I don't think anyone other than you is going to be able to maintain this at present without a lot of research. I guess I will have to wait until you have an interactive working solution and actually trying it out before I can give much more feedback :grin:

**Comment by jahorton** [comment]
> The code LGTM. Not much I can say on the algorithm without spending a very long time on it -- it seems like there is a lot of complexity and that concerns me a little because I don't think anyone other than you is going to be able to maintain this at present without a lot of research. I guess I will have to wait until you have an interactive working solution and actually trying it out before I can give much more feedback 😁

Hmm.  Is there anything in particular that could use additional documentation?  I definitely noted the complexity happening and tried to mitigate it via docs, but it sounds like it's still a bit opaque, based on your feedback.

I'll admit the biggest offender is probably the "pending commit" mechanism - that definitely raised the complexity level, but it's necessary for maintaining segment recognition guarantees once they're made.  "Recognition" is an event (of sorts, even if Promise-based here), so once we've raised that event, we need to commit to it.  Things could behave quite peculiarly if we internally revert the cause of such an event and fire off new events based on new conclusions that ignore the old conclusion.  Performing any sort of 'rollback' would nigh-certainly be far more complex than what we have here.

**Comment by jahorton** [comment]
OK, yeah, the most critical method for the "pending commit" mechanism's integration into the algorithm wasn't exactly documented.  I've just documented that one.

And added a little bit of centralization for split-point management while I was at it, because that was bugging me.

**Comment by mcdurdin** [comment]
I think it's still all a bit of a black box -- as I spend more time on it, it may get better. I worry that the segmentation abstraction that we've come up with may not be the most optimum approach for solving the problem -- it may be overkill and introduce more complexity than we need?

**Pull request merged**